### PR TITLE
bench-tps: Add instruction padding program support

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

In order to double the size of transactions, we need to come up with an appropriate fee model for how much these transactions should cost.

To determine the most appropriate fees, the idea is to compare network TPS when it's only processing large transactions to when it's processing normal-sized transactions.  From that ratio, we can determine a fair fee.

For example, if TPS for normal transactions is 60k, and TPS for 2x transactions is 20k, we can say that 2x transactions cost 3x more, and add that to the normal fee parameters.

To get those TPS numbers, we need bench-tps to interact with a program that can do the same amount of work, but simply have larger transactions.  This way, we isolate the network impact of larger transactions.

#### Summary of Changes

Bench-tps is already a great program for seeing how many transactions the network can process, so let's send some larger transactions at it!

https://github.com/solana-labs/solana-program-library/pull/3608 adds a reference implementation for a program that takes in instruction and account data, and either no-ops, or wraps an inner instruction.  This PR adds support for bench-tps to interact with this program, along with a command-line argument for padding instruction data.

To properly compare apples to apples, you can run this with `--instruction-padding-data-size 0` as the control, and `--instruction-padding-data-size 1800` as the test.

This is a draft because it depends on https://github.com/solana-labs/solana-program-library/pull/3608 -- once that lands, I'll eliminate `inline_instruction_padding_program.rs` and use that crate to create instructions.  Everything else is ready to be reviewed though.

I've tested this with a local validator hacked for larger transactions, and it all works!

To deploy the program, you'll need to run `solana program deploy ./bench-tps/tests/fixtures/spl_instruction_padding.so` and then pass the resulting program id in `--instruction-padding-program-id`.  There are instructions in the CLI failure.

Fixes #

### CODE REVIEW

File: bench-tps/src/cli.rs [hunk 1]
```diff
@@ -456,5 +464,11 @@ pub fn extract_args(matches: &ArgMatches) -> Config {
         args.use_durable_nonce = true;
     }
 
+    args.pad_instruction_data = matches.value_of("pad_instruction_data").map(|s| {
+        s.to_string()
+            .parse()
+            .expect("can't parse padded instruction data size")
```
Comment (lijunwangs) [inline bench-tps/src/cli.rs:470]: can't -> Can't

**Review by lijunwangs — APPROVED** [review]
Looks good

**Comment by KirillLykov** [comment]
The changes looks good: you propagate data to the place where transaction is created.

**Review by KirillLykov — APPROVED** [review]

**Comment by joncinque** [comment]
This is ready for an official look once you have a moment

File: bench-tps/src/bench.rs [hunk 2]
```diff
@@ -601,13 +633,45 @@ fn get_nonce_blockhashes<T: 'static + BenchTpsClient + Send + Sync + ?Sized>(
...
+    let from_pubkey = from_keypair.pubkey();
+    let transfer_instruction = system_instruction::transfer(&from_pubkey, to, lamports);
+    let instruction = if let Some(instruction_padding_config) = instruction_padding_config {
+        create_padded_instruction(
+            instruction_padding_config.program_id,
+            transfer_instruction,
+            vec![],
+            instruction_padding_config.data_size,
+        )
+        .expect("Instruction too large")
```
Comment (lijunwangs) [inline bench-tps/src/bench.rs:654]: How do we know it is for sure instruction too large as the function can fail in multiple ways?

File: bench-tps/src/cli.rs [hunk 3]
```diff
@@ -456,5 +473,19 @@ pub fn extract_args(matches: &ArgMatches) -> Config {
         args.use_durable_nonce = true;
     }
 
+    if let Some(d) = matches.value_of("instruction_padding_data_size") {
```
Comment (lijunwangs) [inline bench-tps/src/cli.rs:476]: d-> data_size makes it easier to read

File: bench-tps/src/inline_instruction_padding_program.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        .len()
+        .try_into()
+        .map_err(|_| ProgramError::InvalidInstructionData)?;
+    data.extend(data_size.to_le_bytes().into_iter());
+    data.extend(instruction.data.into_iter());
+    for i in 0..padding_data {
+        data.push((i % u8::MAX as u32) as u8);
+    }
+
+    // The format for account data goes:
```
Comment (lijunwangs) [inline bench-tps/src/inline_instruction_padding_program.rs:60]: The comment seems does not match the code below. Does it begin with CPI accounts, followed by program account and then padding accounts?

File: bench-tps/src/main.rs [hunk 5]
```diff
@@ -223,6 +224,15 @@ fn main() {
...
         *target_node,
     );
+    if let Some(instruction_padding_config) = instruction_padding_config {
+        info!(
+            "Checking for existence of instruction padding program: {}",
+            instruction_padding_config.program_id
+        );
+        client
+            .get_account(&instruction_padding_config.program_id)
+            .expect("Instruction padding program must be deployed to cluster. Deploy the program using `solana program deploy ./bench-tps/tests/fixtures/spl_instruction_padding.so` and pass the resulting program id with `--instruction-padding-program-id`");
```
Comment (lijunwangs) [inline bench-tps/src/main.rs:234]: --> must be deployed to this cluster.

File: bench-tps/tests/bench_tps.rs [hunk 6]
```diff
@@ -29,8 +33,22 @@ use {
     std::{sync::Arc, time::Duration},
 };
 
+fn program_account(elf: &[u8]) -> AccountSharedData {
```
Comment (lijunwangs) [inline bench-tps/tests/bench_tps.rs:36]: Can you document a little on what is the elf? or maybe rename to elf_code?

File: bench-tps/src/inline_instruction_padding_program.rs (hunk 4 above)
Comment (joncinque) [reply bench-tps/src/inline_instruction_padding_program.rs:60]: Ah yes, you're right, thanks!

File: bench-tps/src/bench.rs (hunk 2 above)
Comment (joncinque) [reply bench-tps/src/bench.rs:654]: Good point, I'll make the error message more generic

File: bench-tps/tests/bench_tps.rs (hunk 6 above)
Comment (joncinque) [reply bench-tps/tests/bench_tps.rs:36]: how about `program_data`?

**Review by lijunwangs — APPROVED** [review]
Looks good to me

**Review by KirillLykov — APPROVED** [review]

**Pull request merged**